Added resource_manager_tags on google_dataproc_cluster resource

mjogdand opened the pull request:
Fixes https://github.com/hashicorp/terraform-provider-google/issues/21918

```release-note: enhancement
dataproc: added `resource_manager_tags` field to `google_dataproc_cluster` resource
```

mjogdand commented:
@roaks3 - Can you please review the PR and advise on the change

roaks3 commented on mmv1/third_party/terraform/services/dataproc/resource_dataproc_cluster.go:
```diff
@@ -670,6 +671,12 @@ func ResourceDataprocCluster() *schema.Resource {
 										Description:  `A map of the Compute Engine metadata entries to add to all instances`,
 									},
 
+									"resource_manager_tags": {
```
Can this field be modified on an existing cluster? If it can, we should include an [update test](https://googlecloudplatform.github.io/magic-modules/test/test/#add-an-update-test). If not, we should add `ForceNew: true`.

mjogdand replied:
You can attach secure tags to a cluster only at the time of cluster creation.
Updation and deletion of secure tags aren't supported.

Reference: https://cloud.google.com/dataproc/docs/guides/use-secure-tags#limitations

Will update schema with `ForceNew: true`

roaks3 commented on mmv1/third_party/terraform/services/dataproc/resource_dataproc_cluster_meta.yaml:
```diff
@@ -27,6 +27,7 @@ fields:
   - field: 'cluster_config.gce_cluster_config.confidential_instance_config.enable_confidential_compute'
   - field: 'cluster_config.gce_cluster_config.internal_ip_only'
   - field: 'cluster_config.gce_cluster_config.metadata'
+  - field: 'cluster_config.gce_cluster_config.resource_manager_tags'
```
Nit: could you move this so the list remains alphabetically sorted

roaks3 commented on mmv1/third_party/terraform/services/dataproc/resource_dataproc_cluster_test.go.tmpl:
```diff
@@ -352,6 +352,33 @@ func TestAccDataprocCluster_withMetadataAndTags(t *testing.T) {
+
+	var cluster dataproc.Cluster
+	randString := acctest.RandString(t, 10)
+	project_id := envvar.GetTestProjectFromEnv()
```
Nit: replace `project_id` with `projectId`

mjogdand replied:
Replaced project_id with projectId

roaks3 commented on mmv1/third_party/terraform/services/dataproc/resource_dataproc_cluster_test.go.tmpl:
```diff
@@ -352,6 +352,33 @@ func TestAccDataprocCluster_withMetadataAndTags(t *testing.T) {
+	var cluster dataproc.Cluster
+	randString := acctest.RandString(t, 10)
+	project_id := envvar.GetTestProjectFromEnv()
+	networkName := acctest.BootstrapSharedTestNetwork(t, "dataproc-cluster")
```
The test ID for these bootstrapped resources should most likely be unique (they are intended to be per-test). Maybe `dataproc-cluster-rmtags` or just `dataproc-cluster2`

mjogdand replied:
Updated test id to dataproc-cluster-rmtags

roaks3 commented on mmv1/third_party/terraform/services/dataproc/resource_dataproc_cluster_test.go.tmpl:
```diff
@@ -1359,6 +1386,70 @@ func validateDataprocCluster_withConfigOverrides(n string, cluster *dataproc.Clu
 	}
 }
 
+func testAccDataprocCluster_withResourceManagerTags(project_id, randString, subnet string) string {
```
Nit: replace `project_id` with `projectId`

mjogdand replied:
Replaced project_id with ProjectId

roaks3 commented on mmv1/third_party/terraform/services/dataproc/resource_dataproc_cluster_test.go.tmpl:
```diff
@@ -1359,6 +1386,70 @@ func validateDataprocCluster_withConfigOverrides(n string, cluster *dataproc.Clu
+func testAccDataprocCluster_withResourceManagerTags(project_id, randString, subnet string) string {
+	return fmt.Sprintf(`
+
+	data "google_project" "project" {
```
Nit: indentation for this function is a bit distracting

mjogdand replied:
Fixed indentation for testing resources

roaks3 reviewed (changes requested):
Mostly LGTM, but left some minor comments, and tests might need to be expanded

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-13490-old..auto-pr-13490) ( 4 files changed, 111 insertions(+), 5 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-13490-old..auto-pr-13490) ( 4 files changed, 111 insertions(+), 5 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 69
Passed tests: 63
Skipped tests: 5
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>dataproc</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccDataprocCluster_withResourceManagerTags</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🔴 Tests failed during RECORDING mode:
`TestAccDataprocCluster_withResourceManagerTags` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cb4894fb-4c36-48b1-bc0e-06ad0233e626/build-log/recording_build/TestAccDataprocCluster_withResourceManagerTags_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cb4894fb-4c36-48b1-bc0e-06ad0233e626/recording/TestAccDataprocCluster_withResourceManagerTags.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cb4894fb-4c36-48b1-bc0e-06ad0233e626/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cb4894fb-4c36-48b1-bc0e-06ad0233e626/recording) for each test

mjogdand commented:
@roaks3 - Can you please share the build logs, I don't have access to view build logs.
If sharing logs is not allowed, Can you please let me know why this VCR test failed so will do required code changes to fix the test.

roaks3 commented:
This is the error we see:

```
    resource_dataproc_cluster_test.go:367: Step 1/1 error: Error running apply: exit status 1
        
        Error: Error waiting for creating Dataproc cluster: Error code 3, message: The resource 'projects/ci-test-project-188019/regions/us-central1/subnetworks/dataproc-cluster-rmtags' is not ready
        
          with google_dataproc_cluster.basic,
          on terraform_plugin_test.tf line 25, in resource "google_dataproc_cluster" "basic":
          25: resource "google_dataproc_cluster" "basic" {
```

I can confirm that the bootstrapped network, subnetwork, and firewall are all created correctly before the dataproc cluster attempts creation, but then we see this error. Because of the way these networks are used for testing, I expect we can get the tests to pass on a re-run, and we will not be able to easily observe this again (because the network resources are persisted).

I'll re-run the tests to see, but could you also briefly check if there is a way to correct for this issue? Mainly, if users go to create this setup, it seems like it would be very easy to hit this issue on their very first run.

roaks3 commented:
/gcbrun

mjogdand commented:
@roaks3 - Thank you for details, I will try to figure out the issue from my end.

Please rerun it once.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-13490-old..auto-pr-13490) ( 4 files changed, 111 insertions(+), 5 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-13490-old..auto-pr-13490) ( 4 files changed, 111 insertions(+), 5 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 69
Passed tests: 63
Skipped tests: 5
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>dataproc</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccDataprocCluster_withResourceManagerTags</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🔴 Tests failed during RECORDING mode:
`TestAccDataprocCluster_withResourceManagerTags` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cb12e92b-a2e0-4f1c-9df6-605874503c91/build-log/recording_build/TestAccDataprocCluster_withResourceManagerTags_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cb12e92b-a2e0-4f1c-9df6-605874503c91/recording/TestAccDataprocCluster_withResourceManagerTags.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cb12e92b-a2e0-4f1c-9df6-605874503c91/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cb12e92b-a2e0-4f1c-9df6-605874503c91/recording) for each test

roaks3 commented:
```
    resource_dataproc_cluster_test.go:367: Step 1/1 error: Error running apply: exit status 1
        
        Error: Error waiting for creating Dataproc cluster: Error code 7, message: Error Code: PERMISSION_DENIED, errorSource: COMPUTE_ENGINE, Error Message: com.google.cloud.resourcemanager.common.error.ExternalStatusException: <eye3 title='PERMISSION_DENIED'/> generic::PERMISSION_DENIED: Permission resourcemanager.tagValueBindings.create denied on resource tagValues/281483646130092 (or it might not exist).
        
        
          with google_dataproc_cluster.basic,
          on terraform_plugin_test.tf line 25, in resource "google_dataproc_cluster" "basic":
          25: resource "google_dataproc_cluster" "basic" {
```

Perhaps we need to wait a small amount of time for the tags to be persisted?

roaks3 reviewed (changes requested):
We'll need to come up with a way to get the tests working

mjogdand commented:
> We'll need to come up with a way to get the tests working

@roaks3 - I trying performing this tests locally and my assumption is that this issue is related to IAM permissions.
Updated the test and added BootstrapIamMembers to grant required roles to compute engine default service account and dataproc service agent service account.

Based on your suggestion added depends_on on cluster resource.
Please review and try to rerun the test.

roaks3 commented on mmv1/third_party/terraform/services/dataproc/resource_dataproc_cluster_test.go.tmpl:
```diff
@@ -352,6 +352,65 @@ func TestAccDataprocCluster_withMetadataAndTags(t *testing.T) {
+	projectId  := envvar.GetTestProjectFromEnv()
+
+	acctest.BootstrapIamMembers(t, []acctest.IamMember{
+        {
```
Nit: could you clean up the indentation for this block?

mjogdand replied:
Formatted the code

roaks3 commented on mmv1/third_party/terraform/services/dataproc/resource_dataproc_cluster_test.go.tmpl:
```diff
@@ -1359,6 +1418,76 @@ func validateDataprocCluster_withConfigOverrides(n string, cluster *dataproc.Clu
+  }
+
+
+}
```
This looks like an extra curly brace? Cleaning up indentation might help catch something like this.

mjogdand replied:
Removed a extra curly brace and re-formatted the code.
Thank You

roaks3 reviewed (changes requested):
Fix seems reasonable. I don't think the test can be run in the current state, but once updated I'll kick that off to see.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-13490-old..auto-pr-13490) ( 4 files changed, 146 insertions(+), 5 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-13490-old..auto-pr-13490) ( 4 files changed, 146 insertions(+), 5 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 69
Passed tests: 63
Skipped tests: 5
Affected tests: 1

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>dataproc</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccDataprocCluster_withResourceManagerTags</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/develop/test/test/)

modular-magician commented:
🔴 Tests failed during RECORDING mode:
`TestAccDataprocCluster_withResourceManagerTags` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cde850dd-d17a-4bcd-802e-c53d93741f8a/build-log/recording_build/TestAccDataprocCluster_withResourceManagerTags_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cde850dd-d17a-4bcd-802e-c53d93741f8a/recording/TestAccDataprocCluster_withResourceManagerTags.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cde850dd-d17a-4bcd-802e-c53d93741f8a/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-13490/artifacts/cde850dd-d17a-4bcd-802e-c53d93741f8a/recording) for each test
